File: recycleradapter/src/main/java/com/github/nukc/recycleradapter/BaseAdapter.java
package com.github.nukc.recycleradapter;

import android.support.annotation.LayoutRes;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Nukc.
 */

abstract class BaseAdapter<T> extends RecyclerView.Adapter<RecyclerHolder> {

    protected List<T> mDataList;

    public BaseAdapter() {
        mDataList = new ArrayList<>(0);
    }

    public BaseAdapter(List<T> dataList) {
        mDataList = dataList;
    }

    @Override
    public int getItemCount() {
        return mDataList.size();
    }

    public View inflate(ViewGroup parent, @LayoutRes int resource) {
        return LayoutInflater.from(parent.getContext()).inflate(resource, parent, false);
    }

    public T getItem(int position) {
        return mDataList.get(position);
    }

    public List<T> getDataList() {
        return mDataList;
    }

    public void refresh(List<T> dataList) {
        mDataList.clear();
        mDataList.addAll(dataList);
        notifyDataSetChanged();
    }

    public void add(int position, T date) {
        mDataList.add(position, date);
        notifyItemInserted(position);
    }

    public void add(T data) {
        add(mDataList.size(), data);
    }

    public void addAll(int positionStart, List<T> dataList) {
        mDataList.addAll(positionStart, dataList);
        notifyItemRangeInserted(positionStart, dataList.size());
    }

    public void addAll(List<T> dataList) {
        addAll(mDataList.size(), dataList);
    }

    public void move(int fromPosition, int toPosition) {
        T remove = mDataList.remove(fromPosition);
        mDataList.add(toPosition, remove);
        notifyItemMoved(fromPosition, toPosition);
    }

    public void change(int position, T data) {
        mDataList.remove(position);
        mDataList.add(position, data);
        notifyItemChanged(position);
    }

    public void remove(int position) {
        mDataList.remove(position);
        notifyItemRemoved(position);
    }

    public void clear() {
        mDataList.clear();
        notifyDataSetChanged();
    }
}

File: recycleradapter/src/main/java/com/github/nukc/recycleradapter/BaseAdapter.kt
package com.github.nukc.recycleradapter

import androidx.annotation.MainThread
import androidx.recyclerview.widget.RecyclerView

abstract class BaseAdapter(val items: MutableList<Any>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    override fun getItemCount(): Int {
        return items.size
    }

    @MainThread
    fun refresh(items: MutableList<Any>) {
        this.items.clear()
        this.items.addAll(items)
        notifyDataSetChanged()
    }

    @MainThread
    fun addAll(items: MutableList<Any>) {
        addAll(this.itemCount, items)
    }

    @MainThread
    fun addAll(positionStart: Int, items: MutableList<Any>) {
        this.items.addAll(positionStart, items)
        notifyItemRangeInserted(positionStart, items.size)
    }

    @MainThread
    fun add(position: Int, item: Any) {
        items.add(position, item)
        notifyItemInserted(position)
    }

    @MainThread
    fun add(item: Any) {
        add(items.size, item)
    }

    @MainThread
    fun change(position: Int, item: Any) {
        items[position] = item
        notifyItemChanged(position)
    }

    @MainThread
    fun remove(position: Int) {
        items.removeAt(position)
        notifyItemRemoved(position)
    }

    @MainThread
    fun move(fromPosition: Int, toPosition: Int) {
        val item = items.removeAt(fromPosition)
        items.add(toPosition, item)
        notifyItemMoved(fromPosition, toPosition)
    }

    @MainThread
    fun clear() {
        items.clear()
        notifyDataSetChanged()
    }
}